pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,282 @@
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// `spec` is the specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// `status` is the current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// `metadata` is the standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// `items` is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// `priorityLevelConfiguration` should reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfiguration PriorityLevelConfigurationReference
+	// `matchingPrecedence` is used to choose among the FlowSchemas that match a given request.  The chosen
+	// FlowSchema is among those with the numerically lowest (which we take to be logically highest)
+	// MatchingPrecedence.  Each MatchingPrecedence value must be non-negative.
+	MatchingPrecedence int32
+	// `distinguisherMethod` defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	DistinguisherMethod *FlowDistinguisherMethod
+	// `rules` describes which requests will match this flow schema.
+	Rules []PolicyRuleWithSubjects
+}
+
+// FlowDistinguisherMethodType is the type of flow distinguisher method
+type FlowDistinguisherMethodType string
+
+// FlowDistinguisherMethod specifies the method of a flow distinguisher.
+type FlowDistinguisherMethod struct {
+	// `type` is the type of flow distinguisher method
+	Type FlowDistinguisherMethodType
+}
+
+// PriorityLevelConfigurationReference contains information that points to the "request-priority" being used.
+type PriorityLevelConfigurationReference struct {
+	// `name` is the name of resource being referenced
+	Name string
+}
+
+// PolicyRuleWithSubjects prescribes a test that applies to a request to an apiserver. The test considers the subject
+// making the request, the verb being requested, and the resource to be acted upon.
+type PolicyRuleWithSubjects struct {
```

Comment:
By comparison with rbac, I have realized that we are missing something here.  I asked myself why does rbac have ClusterRole as well as Role?  The answer is about the namespace of the object that the request accesses.  Note that PolicyRuleWithSubjects currently says nothing about this namespace.

We could add ClusterPolicyRule and have to deal with unions.  But I think it would be better to add `Namespace` here or to `PolicyRule` and stipulate that the Namespace is empty for a PolicyRule that references non-namespaces resources and non-resources.

Reply:
yeah, +1 .. how did i miss that in the first place..

Reply:
but wait.. but do we really need a flow-schema to match a set of namespace? can we achieve the same effect using the by-namespace flow-distinguisher?

Reply:
I just noticed that I forgot to say that we should define a value of the `Namespace` field that, in the case of namespaced objects, does not restrict on namespace (e.g., `NamespaceAll = "*"` or `NamespaceAll = ""`).

Remember that one FlowSchema has a slice of Subject, so it is enough for each Subject to be specific to one namespace (or not filter on namespace) --- unless and until we discover a use case where what is desired is something like a regex on namespace (and we have _that_ issue on all our fields, again this is one reason I started with a predicate-based approach).

Reply:
Note also that the flow distinguisher method plays no role in deciding whether a given FlowSchema matches a given request.  So no, we can not use any flow distinguisher method to get the effect of restricting a FlowSchema to match only requests from service accounts in a given namespace.

Reply:
Note that the example configuration in the KEP (https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#example-configuration) includes some filtering on the object namespace.